### ISSUE TITLE
Tab Missing VALUE

### ISSUE THREAD

**Issue by MSDP** [docusign/DocuSign-.NET-Client#3]
It is not possible to pre-fill textFields without Value. 

Should be added to Tab: 

public class Tab
	{
        /// <summary>
        /// Tab Value
        /// </summary>
        public string value { get; set; }

**Comment by InbarGazit** [comment]
Thanks, we'll add this to our next version. Stay tuned...

**Comment by MSDP** [comment]
I added it to a fork of my own, and it wasn't functional unfortunately. I am still unable to add a value to a Data Field on a Template.

**Comment by MSDP** [comment]
Unless there is another method to add Values to Data Fields on a Template that I am unaware of?

**Comment by peterlord** [comment]
Also facing exactly the same problem. I too tried to clone and build in VS but missing the CustomConmfig.csproj and therefore having trouble building the project. Any ideas?

**Comment by InbarGazit** [comment]
@MSDP , here is an example of a REST call to set value of TAB. We will support it in a later version of the nuget package

"emailBlurb": "overide default blurb for via json",
"emailSubject": "test",
"status": "created",
"templateId": "e14fe28c-007e-44d2-8485-275fe134f43c",
"templateRoles": [{
"email": "t@test.com",
"name": "Fat Albert",
"roleName": "Signer",
"tabs": {
"radioGroupTabs": [{
"groupName": "parent",
"radios": [{
"value&q
uot;: "no",
					"selected": true
				}]
			}]
		}
	}]
}

**Comment by MSDP** [comment]
I added Value to the Tab class, shouldn't it just include it?

On Fri, Jan 23, 2015 at 2:05 PM, Inbar Gazit <notifications@github.com>
wrote:

> @MSDP <https://github.com/MSDP> , here is an example of a REST call to
> set value of TAB. We will support it in a later version of the nuget package
>
> "emailBlurb": "overide default blurb for via json",
> "emailSubject": "test",
> "status": "created",
> "templateId": "e14fe28c-007e-44d2-8485-275fe134f43c",
> "templateRoles": [{
> "email": "t@test.com",
> "name": "Fat Albert",
> "roleName": "Signer",
> "tabs": {
> "radioGroupTabs": [{
> "groupName": "parent",
> "radios": [{
> "value&q
> uot;: "no",
> "selected": true
> }]
> }]
> }
> }]
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docusign/DocuSign-.NET-Client/issues/3#issuecomment-71246465>
> .
>



-- 


*Adam Reed*
*(P) 716 478 3509*
*(C) 716 406 7333*
*Chief Technical Officer*
*Spire Recovery Solutions, LLC*
*Tier One Asset Management Group, LLC *
*Market Street Debt Partners, LLC.*

**Comment by InbarGazit** [comment]
@MSDP , can you tell me a bit more? what exactly are you doing? what type of tab is it?

**Comment by InbarGazit** [comment]
@peterlord , your issue has to do with the unit test project. I just pushed an update that fixes these things. you do not need CustomConfig. You do need (if you want to run the unit tests) to update the values in ConfigLoader.cs to what ever values you're using

**Comment by MSDP** [comment]
var tabCollection = new TabCollection();
            var fileNumber = new Tab()
            {
                tabLabel = "FileNumber",
                value = "Test",
                xPosition = 645,
                yPosition = 309,
                documentId = 1,
                pageNumber = 1
            };
            tabCollection.textTabs.Add(fileNumber);

            // create envelope object and assign login info
            var envelope = new Envelope
            {
                Login = account,
                TemplateId = "********",
                TemplateRoles = new TemplateRole[]
                {
                    new TemplateRole()
                    {
                        email = RecipientEmail,
                        name = RecipientName,
                        roleName = "Consumer",
                        routingOrder = "1"
                    }
                },
                EmailSubject = "DocuSign .NET Client - Signature Request from Template",
                Status = "sent"
            };
            envelope.AddTabs(tabCollection);

**Comment by InbarGazit** [comment]
@MSDP , the C# code currently does not support value. You can try to add that support yourself, but simply adding string value {get; set;} is not going to be enough. Here are examples of the REST calls  (both GET and POST) needed to set a value on a textTab:

GET http://localhost/restapi/vdev/accounts/d51cfe40-7b1b-4369-8e87-c2b62eeba9de/templates/153BF6A6-0ABE-40B0-80AC-9BAA7E5CE02F/recipients/1/tabs
X-DocuSign-Authentication: <DocuSignCredentials><Username>acostasy@me.com</Username><Password>19868899</Password><IntegratorKey>DOCU-190f53a1-4615-4928-9709-5aa9cc3c60d3</IntegratorKey></DocuSignCredentials>
Accept: application/json; charset=utf-8
Content-Type: application/json; charset=utf-8
X-DocuSign-ClientTransactionId: T635576188451436406
X-DocuSign-PrettyPrint: true

200 OK
Pragma: no-cache
X-DocuSign-ClientTransactionId: T635576188451436406
Content-Length: 777
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Fri, 23 Jan 2015 22:07:25 GMT
Expires: -1

{
  "textTabs": [
    {
      "validationPattern": "",
      "validationMessage": "",
      "shared": "false",
      "requireInitialOnSharedChange": "false",
      "name": "Test A",
      "value": "ABCD",
      "required": "false",
      "locked": "false",
      "concealValueOnDocument": "false",
      "disableAutoSize": "true",
      "tabLabel": "AccountTitle 122",
      "bold": "false",
      "italic": "false",
      "underline": "false",
      "documentId": "1",
      "recipientId": "1",
      "pageNumber": "1",
      "xPosition": "312",
      "yPosition": "62",
      "width": "60",
      "height": "22",
      "tabId": "6049a9d0-32ed-4522-979c-209bc0e02cb2"
    }
  ]
}

============ NEXT REQUEST ===============

POST http://localhost/restapi/vdev/accounts/d51cfe40-7b1b-4369-8e87-c2b62eeba9de/envelopes
X-DocuSign-Authentication: <DocuSignCredentials><Username>acostasy@me.com</Username><Password>19868899</Password><IntegratorKey>DOCU-190f53a1-4615-4928-9709-5aa9cc3c60d3</IntegratorKey></DocuSignCredentials>
Accept: application/json; charset=utf-8
Content-Type: application/json; charset=utf-8
X-DocuSign-ClientTransactionId: T635576188457356998
X-DocuSign-PrettyPrint: true
{
  "templateId": "153BF6A6-0ABE-40B0-80AC-9BAA7E5CE02F",
  "templateRoles": [
    {
      "email": "acostays@me.company",
      "roleName": "Signer",
      "name": "test A",
      "tabs": {
        "textTabs": [
          {
            "value": "default",
            "tabLabel": "AccountTitle 122",
            "documentId": "1",
            "recipientId": "1",
            "pageNumber": "1",
            "xPosition": "100",
            "yPosition": "100"
          }
        ]
      }
    }
  ],
  "status": "sent",
  "emailSubject": "This is a test"
}
201 Created
Pragma: no-cache
X-DocuSign-ClientTransactionId: T635576188457356998
Content-Length: 195
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Fri, 23 Jan 2015 22:07:30 GMT
Expires: -1

{
  "envelopeId": "442cfbc2-06b5-4b5b-a05c-b4953cb6adb4",
  "uri": "/envelopes/442cfbc2-06b5-4b5b-a05c-b4953cb6adb4",
  "statusDateTime": "2015-01-24T15:07:29.5200000Z",
  "status": "sent"
}

============ NEXT REQUEST ===============

GET http://localhost/restapi/vdev/accounts/d51cfe40-7b1b-4369-8e87-c2b62eeba9de/envelopes/442cfbc2-06b5-4b5b-a05c-b4953cb6adb4/recipients/1/tabs
X-DocuSign-Authentication: <DocuSignCredentials><Username>acostasy@me.com</Username><Password>19868899</Password><IntegratorKey>DOCU-190f53a1-4615-4928-9709-5aa9cc3c60d3</IntegratorKey></DocuSignCredentials>
Accept: application/json; charset=utf-8
Content-Type: application/json; charset=utf-8
X-DocuSign-ClientTransactionId: T635576188504911753
X-DocuSign-PrettyPrint: true

200 OK
Pragma: no-cache
X-DocuSign-ClientTransactionId: T635576188504911753
Content-Length: 701
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Fri, 23 Jan 2015 22:07:30 GMT
Expires: -1

{
  "textTabs": [
    {
      "validationPattern": "",
      "validationMessage": "",
      "shared": "false",
      "requireInitialOnSharedChange": "false",
      "name": "Test A",
      "value": "default",
      "required": "false",
      "locked": "false",
      "concealValueOnDocument": "false",
      "disableAutoSize": "true",
      "tabLabel": "AccountTitle 122",
      "documentId": "1",
      "recipientId": "1",
      "pageNumber": "1",
      "xPosition": "312",
      "yPosition": "62",
      "width": "60",
      "height": "22",
      "tabId": "d9872110-65a6-4d2e-9943-9308da161aea"
    }
  ]
}

**Comment by InbarGazit** [comment]
Basically, what you would have to do is add tabs to the initial call creating the envelope. Calling AddTabs() after an envelope was already created would not update the value.

**Comment by MSDP** [comment]
@InbarGazit So Tabs/TabCollection are not functional at all?

**Comment by MSDP** [comment]
@InbarGazit I call AddTabs before I create the Envelope? 

var tabCollection = new TabCollection();
            var fileNumber = new Tab()
            {
                tabLabel = "FileNumber",
                value = "Test",
                xPosition = 645,
                yPosition = 309,
                documentId = 1,
                pageNumber = 1
            };
            tabCollection.textTabs.Add(fileNumber);

            // create envelope object and assign login info
            var envelope = new Envelope
            {
                Login = account,
                TemplateId = "B6865402-15F0-4E8B-BA03-655B1DC210BA",
                TemplateRoles = new TemplateRole[]
                {
                    new TemplateRole()
                    {
                        email = RecipientEmail,
                        name = RecipientName,
                        roleName = "Consumer",
                        routingOrder = "1"
                    }
                },
                EmailSubject = "DocuSign .NET Client - Signature Request from Template",
                Status = "sent"
            };
            envelope.AddTabs(tabCollection);
            result = envelope.Create();

**Comment by InbarGazit** [comment]
@MSDP it all depends on what you're doing. When you're using a template - the tabs are already defined and you cannot use a POST method you need to use a PUT if you wish to update them. This type of call (UpdateTab) does not yet exist in this C# library. 
The way to use tabs with non-template envelope is to add them to the Recipient Object when you create an envelope. so you create your TabCollection object and then assign recipient.tabs to it. Then you call the create() method to create a new envelope where the recipient above is part of the Recipients object on the Envelope .hope that makes sense, if not - I can try to provide some code sample

**Comment by MSDP** [comment]
Inbar, I understand, I was hoping more of this was included in the Helper Library. I see there isn't any functionality for adding Tabs to Recipient Objects either unfortunately. 

Is there an ETA on the new release that will include this functionality?

**Comment by MSDP** [comment]
Inbar,

Looking closer at your request order. Can I use the Client Library to create the Envelope using the Template, and set it as "created" to make it a Draft, then call: 

POST http://DOCUSIGN/restapi/v2/accounts/accountid/envelopes

with the Tab Info, then Update the Envelop to Sent?

**Comment by InbarGazit** [comment]
well, you can do it, but not using the library. The library's support for adding tabs is only for non-template envelopes, adding the feature you asked for templates is something we haven't yet considered. If you write the code yourself and you have it tested and working and are willing to share it - we may be able to include your addition quicker.

**Comment by MSDP** [comment]
Inbar, I understand, I would like to create an envelope from template using the Library as a draft, then Update that envelope using a REST Request. 

Could you direct me to the proper request for updating a created envelope?

**Comment by InbarGazit** [comment]
are you talking about a template or a regular envelope?
if the latter - then you can first call the Create() method and then you can call a variety of things like Add/Remove documents, AddRecipients, UpdateStatus() etc. As for the AddTabs() method, it would only work if you already added the recipient and has the recipientId information for that recipient in the AddTabs call

**Comment by InbarGazit** [comment]
note that when you call AddRecipients, you pass in a recipient object that may include tabs. Same goes for when you create the envelope initially (using one of the overloads) so that may be the better way to get tabs (if you can do that in one call)

**Comment by InbarGazit** [comment]
@MSDP , one thing, AddTabs() only works on the first recipient in the envelope. that's not good enough, I know. We'll add another overload of it that takes a recipient ID. Note this part:

                req.Uri = string.Format("/envelopes/{0}/recipients/{1}/tabs", EnvelopeId, (string)(GetFirstRecipients().First())["recipientIdGuid"]);

See that we grab the first recipient and use the recipientIdGuid property for the REST call. This means you cannot use this for any recipient. It would be real easy to add another overload that takes this guid as a parameter. To find the guid, you would have to also add a way to do that. Note in GetFirstRecipients() that it makes a GET call and then just returns some of the recipients (minimum routing order). you might need to change this to get all recipients and find the one you need for your case. 
Again, it's much easier to do that in a single call - both add recipients and tabs for these recipients.

**Comment by MSDP** [comment]
Inbar, 

Considering your suggestion, will this not function?



         bool result = account.Login();

            var recipientName = "John Doe";
            var recipientEmail = "support@****";


            if (!result)
            {
                Console.WriteLine("Login API call failed for user {0}.\nError Code:  {1}\nMessage:  {2}", account.Email,
                    account.RestError.errorCode, account.RestError.message);
                return;
            }

            var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 645,
                yPosition = 309,
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1,
                
            });

            var envelope = new Envelope
            {
                Login = account,
                EmailSubject = "DocuSign .NET Client - Signature Request from Template",
                Status = "created",
                Recipients = new Recipients()
                {
                    signers = new Signer[]
                    {
                        new Signer()
                        {
                            email = recipientEmail,
                            name = recipientName,
                            routingOrder = "1",
                            recipientId = "1",
                            tabs = new Tabs { textTabs = tabCollection.textTabs.ToArray() }
                        }
                    }
                }
            };

            result = envelope.Create("PATH");

            if (!result)
            {
                if (envelope.RestError != null)
                {
                    Console.WriteLine("Error code:  {0}\nMessage:  {1}", envelope.RestError.errorCode,
                        envelope.RestError.message);
                    return;
                }
                else
                {
                    Console.WriteLine(
                        "Error encountered while requesting signature from template, please review your envelope and recipient data.");
                    return;
                }
            }
            else
            {
                Console.WriteLine("Signature request has been sent to {0}, envelopeId is {1}",
                    envelope.TemplateRoles[0].email, envelope.EnvelopeId);
            }
            Console.ReadLine();

**Comment by MSDP** [comment]
Guess Not 



Error code:  TAB_REFERS_TO_MISSING_DOCUMENT
Message:  The DocumentId specified in the tab element does not refer to a docume
nt in this envelope. Tab refers to DocumentId 0 which is not present.

**Comment by MSDP** [comment]
Re-Reading your response:

I need to Create the Envelope in Draft Mode
Call AddDocuments
Call AddRecipents
Call AddTabs

Then UpdateStatus?

**Comment by MSDP** [comment]
I was able to create the Envelope, attach my document, and add my Recipents. 

However, when reviewing the document, the Tab does not show on the document. 

Code:

            result = envelope.AddRecipients(new Recipients()
            {
                signers = new Signer[]
                {
                    new Signer()
                    {
                        email = recipientEmail,
                        name = recipientName,
                        routingOrder = "1",
                        recipientId = "1",
                        tabs = new Tabs {textTabs = tabCollection.textTabs.ToArray()}
                    }
                }
            });

**Comment by InbarGazit** [comment]
if this is all the code - textTabs is empty. Where is the code with the tab itself?

**Comment by MSDP** [comment]
Inbar,

That was just my snippet, see textTabs code here:

            var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 645,
                yPosition = 309,
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1
                
            });

**Comment by MSDP** [comment]
I also called AddTabs after AddRecipents and still found nothing on the document. 


var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 645,
                yPosition = 309,
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1
                
            });
            tabCollection.signHereTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1192,
                pageNumber = 1
            });
            tabCollection.fullNameTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1171,
                pageNumber = 1
            });
            tabCollection.dateSignedTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1295,
                pageNumber = 1
            });

            result = envelope.AddTabs(tabCollection);
            if (!result)
            {
                if (envelope.RestError != null)
                {
                    Console.WriteLine("Error code:  {0}\nMessage:  {1}", envelope.RestError.errorCode, envelope.RestError.message);
                    return;
                }
                else
                {
                    Console.WriteLine("Error encountered while requesting signature from template, please review your envelope and recipient data.");
                    return;
                }
            }
            else
            {
                Console.WriteLine("Tabs Added");
            }

**Comment by InbarGazit** [comment]
@MSDP , I'm sorry, it appears for now you might need to do it all in one single call. I'll have to figure out why later, but to get you going, here is an example of code that works:

            // create empty envelope
            var envelope = new Envelope{ Login = //TODO - add your user};
            byte[] bytes = { 36, 45, 34, 67, 121, 87, 99, 32, 32, 32, 54, 54, 55, 56, 32 };
            var recp = new Recipients();
            var signers = new List<Signer>();
            var signer = new Signer();
            signer.name = "test test2";
            signer.email = "test5@test2.com";
            signer.routingOrder = "1";
            signer.recipientId = "1";
            var tabs = new List<Tab>();
            var tab = new Tab();
            tab.xPosition = 10;
            tab.yPosition = 10;
            tab.required = true;
            tab.fontSize = "Size10";
            tab.width = 100;
            tab.height = 20;
            tab.pageNumber = 1;
            tab.recipientId = "1";
            tab.documentId = 1;
            tab.tabLabel = "test";
            tabs.Add(tab);
            signer.tabs = new Tabs();
            signer.tabs.signHereTabs = tabs.ToArray();
            signers.Add(signer);
            recp.signers = signers.ToArray();
            envelope.Recipients = recp;
            envelope.EmailSubject = "test-new";
            envelope.Create(bytes, "test1.doc");

**Comment by MSDP** [comment]
Error code:  TAB_REFERS_TO_MISSING_DOCUMENT
Message:  The DocumentId specified in the tab element does not refer to a docume
nt in this envelope. Tab refers to DocumentId 0 which is not present.


            bool result = account.Login();

            var recipientName = "John Doe";
            var recipientEmail = "support@ms-dp.net";


            if (!result)
            {
                Console.WriteLine("Login API call failed for user {0}.\nError Code:  {1}\nMessage:  {2}", account.Email,
                    account.RestError.errorCode, account.RestError.message);
                return;
            }

            var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 645,
                yPosition = 309,
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1
                
            });
            tabCollection.signHereTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1192,
                pageNumber = 1
            });
            tabCollection.fullNameTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1171,
                pageNumber = 1
            });
            tabCollection.dateSignedTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1295,
                pageNumber = 1
            });


            var envelope = new Envelope
            {
                Login = account,
                EmailSubject = "DocuSign .NET Client - Signature Request from Template",
                Status = "created",
            };


            var recp = new Recipients();
            var signers = new List<Signer>();
            var signer = new Signer
            {
                name = recipientName,
                email = recipientEmail,
                routingOrder = "1",
                recipientId = "1",
                tabs = new Tabs
                {
                    signHereTabs = tabCollection.signHereTabs.ToArray(),
                    textTabs = tabCollection.textTabs.ToArray(),
                    dateSignedTabs = tabCollection.dateSignedTabs.ToArray(),
                    fullNameTabs = tabCollection.fullNameTabs.ToArray()
                }
            };
            signers.Add(signer);
            recp.signers = signers.ToArray();
            envelope.Recipients = recp;
            result = envelope.Create(@"C:\Users\adam_000\Documents\[SBS]_Release_and_Authorization_Form.pdf");
            if (!result)
            {
                if (envelope.RestError != null)
                {
                    Console.WriteLine("Error code:  {0}\nMessage:  {1}", envelope.RestError.errorCode, envelope.RestError.message);
                    return;
                }
                else
                {
                    Console.WriteLine("Error encountered while requesting signature from template, please review your envelope and recipient data.");
                    return;
                }
            }
            else
            {
                Console.WriteLine("Signature request has been sent to {0}, envelopeId is {1}",
                    envelope.Recipients.signers.First(), envelope.EnvelopeId);
            }

**Comment by InbarGazit** [comment]
@MSDP  this is good news. We know exactly what the issue is. You either didn't add a document at all, or your document ID is different and all you have to do is change the tab.DocumentId property to match what you set when you added the document.

**Comment by MSDP** [comment]
I believe it is the Create Function? 

You used: Create(bytes, FileName);
     I used: Create(FileLocation);

**Comment by InbarGazit** [comment]
try to specify documentId = 1 in the tab object and see if this does the trick

**Comment by MSDP** [comment]
Error code:  TAB_OUT_OF_BOUNDS
Message:  Tab is placed off of the page. Tab "SignHere" is located off of page 1

I got the X and Y from building a Template and pulling the Location. 

            var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 645,
                yPosition = 309,
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.signHereTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1192,
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.fullNameTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1171,
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.dateSignedTabs.Add(new Tab
            {
                xPosition = 404,
                yPosition = 1295,
                pageNumber = 1,
                documentId = 1
            });

**Comment by InbarGazit** [comment]
@MSDP , these are excellent error reports, we know exactly what's wrong. not sure how tall is your page, but apparently 1192 pixels is too far outside. Just to try, set this to a very low number and see what happens....

**Comment by MSDP** [comment]
I set 200 x 200 and it worked!!!! =D 

However, even tho I set 200x200, if I click into the Editor on the website, it shows it is now at: 415x459. 

The measurements don't seem to be matching up?

**Comment by InbarGazit** [comment]
@MSDP, would you mine sharing the full latest code you have? (minus any identifiers you can remove)

I think there are some issues with pixels conversions that I need to look at, but you might be able to play around with the numbers and figure out how it's working...

**Comment by MSDP** [comment]
Sure!

The document in question is in my account: adam@ms-dp.net if you have access to review it. 

          bool result = account.Login();

            var recipientName = "John Doe";
            var recipientEmail = "support@ms-dp.net";


            if (!result)
            {
                Console.WriteLine("Login API call failed for user {0}.\nError Code:  {1}\nMessage:  {2}", account.Email,
                    account.RestError.errorCode, account.RestError.message);
                return;
            }

            var tabCollection = new TabCollection();
            tabCollection.textTabs.Add(new Tab
            {
                xPosition = 200, //645
                yPosition = 200, //309
                tabLabel = "FileNumber",
                value = "TESTVALUE",
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.signHereTabs.Add(new Tab
            {
                xPosition = 200, //404
                yPosition = 200, //1192
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.fullNameTabs.Add(new Tab
            {
                xPosition = 200, //404
                yPosition = 200, //1171
                pageNumber = 1,
                documentId = 1
            });
            tabCollection.dateSignedTabs.Add(new Tab
            {
                xPosition = 200, //404
                yPosition = 200, //1295
                pageNumber = 1,
                documentId = 1
            });
            
            var envelope = new Envelope
            {
                Login = account,
                EmailSubject = "DocuSign .NET Client - Signature Request from Template",
                Status = "created",
            };
            
            var recp = new Recipients();
            var signers = new List<Signer>();
            var signer = new Signer
            {
                name = recipientName,
                email = recipientEmail,
                routingOrder = "1",
                recipientId = "1",
                tabs = new Tabs
                {
                    signHereTabs = tabCollection.signHereTabs.ToArray(),
                    textTabs = tabCollection.textTabs.ToArray(),
                    dateSignedTabs = tabCollection.dateSignedTabs.ToArray(),
                    fullNameTabs = tabCollection.fullNameTabs.ToArray()
                }
            };
            signers.Add(signer);
            recp.signers = signers.ToArray();
            envelope.Recipients = recp;

            result = envelope.Create(@"C:\Users\adam_000\Documents\[SBS]_Release_and_Authorization_Form.pdf");

            if (!result)
            {
                if (envelope.RestError != null)
                {
                    Console.WriteLine("Error code:  {0}\nMessage:  {1}", envelope.RestError.errorCode, envelope.RestError.message);
                    return;
                }
                else
                {
                    Console.WriteLine("Error encountered while requesting signature from template, please review your envelope and recipient data.");
                    return;
                }
            }
            else
            {
                Console.WriteLine("Signature request has been sent to {0}, envelopeId is {1}",
                    envelope.Recipients.signers.First(), envelope.EnvelopeId);
            }
            Console.ReadLine();

**Comment by MSDP** [comment]
xPosition = 145, //
yPosition = 109, //309

Turns into
From left: 301  From top: 226

**Comment by MSDP** [comment]
Halving the numbers shown on the Web Designer, ALMOST perfectly align. Seems your code is doubling the positioning.

**Comment by MSDP** [comment]
@InbarGazit 

I was able to get it working by meticulously changing the values then viewing them in the Web Designer! Hurray! Now I must ask, how can I set the Envelope Settings?

Envelope Settings
Recipient Information
Allow recipient(s) to change signing responsibility.
Reminders and Expirations
Envelope will expire on 2/22/2015.
Signer will be warned of expiration on 2/17/2015
Signer will be reminded on 1/24/2015.
Then, every day after that.

Reminders Enabled
Send a reminder email to the signer inday(s).
After the first reminder, send a reminder everyday(s).
Expire envelope in day(s).
Warn signer of envelope expirationday(s)

**Comment by InbarGazit** [comment]
see the Envelope class has customFields property - that's how you add settings. Again, I believe right now you would have to do that at the same time you create the envelope, we can add the ability to update them separately later...

**Comment by InbarGazit** [comment]
@MSDP  , wanted to also personally invite you to our very first developer conference to be held this coming March in San Francisco. this is exactly the opportunity to meet me and other developers in DocuSign and talk shop. And the cost is very low. See all the information here http://momentum.docusign.com/san-francisco/dev-con/?mc=devcon-github

**Comment by MSDP** [comment]
@InbarGazit Thank you, I wish I could attend, unfortunately a trip from NY to San Fran is not in our companies budget :(. 

I got everything working by slowly tweaking the locations, but the provided X/Y shows up on the form as double the X/Y provided.

**Comment by InbarGazit** [comment]
@MSDP , sorry it took me a while to track this down. the confusion is over DPI/pixels. The formulas are (x * 150/72), and (x * 72 / 150) which comes out to be roughly double & half but if you want to be exact - use the conversion above. We would work to better document this so the confusion is minimized for future developers. Thanks for being a pinner and help us be better!

**Comment by MSDP** [comment]
Inbar,

I try my best! So let me ask you this, what is the formula to convert the
Locations shown in the Web Editor to API numbers.

The reverse of (x * 150/72) ?

**Comment by InbarGazit** [comment]
yup, you got it

**Comment by jeff-ro** [comment]
Hi Inbar,

I had the same issue, trying to set value of textTabs in a template. I just got the latest DocuSign.Integrations.Client project, but it's not compiling w/ the error "Source file 'TextTab.cs' could not be found".  Although referenced by the .csproj, it is not included in the project.

Thanks!

**Comment by InbarGazit** [comment]
@jeff-ro for now see if you can use this https://raw.githubusercontent.com/gonemad83/DocuSign-.NET-Client/master/DocuSign.Integrations.Client/TextTab.cs
I'm waiting for @gonemad83 to provide the file as part of a branch we can merge in, if I don't get it in the next 2 days - we'll add it ourselves

**Comment by jeff-ro** [comment]
Worked like a charm, many thanks for the quick response ~

**Comment by InbarGazit** [comment]
@MSDP  the original issue here was fixed with latest code and I'm checking with you if I can close this issue?

**Comment by tranquilChaos** [comment]
@InbarGazit  - is this supported yet? I just updated to 1.5.1 and still only see the option for textTabs under Envelope.TemplateRoles[0].tabs ... I'm kind of in a tough spot right now because I need to be able to populate a template with radio groups, checkboxes and number tabs, but the .Net Client API doesn't support it from what I can tell. If I try to use the code sample you provide above with signers and recipients, I get the error: recipients may not be specifed when templateId is set. Use templateRoles. So basically, I'm forced to use TemplateRoles if I use a template and the API only supports text tabs. Am I missing something?

here is an example of a REST call to set value of TAB. We will support it in a later version of the nuget package

"emailBlurb": "overide default blurb for via json",
"emailSubject": "test",
"status": "created",
"templateId": "e14fe28c-007e-44d2-8485-275fe134f43c",
"templateRoles": [{
"email": "t@test.com",
"name": "Fat Albert",
"roleName": "Signer",
"tabs": {
"radioGroupTabs": [{
"groupName": "parent",
"radios": [{
"value&q
uot;: "no",
"selected": true
}]
}]
}
}]
}

**Comment by tranquilChaos** [comment]
Actually, I forked and modified the client to support this. Basically, I added support for CheckboxTab, NumberTab and RadioGroupTab (the three I needed) to TemplateRole. You guys already had full support for it in the RequestBuilder and all I needed to do was create the property classes, uncomment and re-type in RoleTabs and everything worked fine. We are currently using this in our app. I didn't add any unit tests.
